File: python/examples/timing_example.py
"""Example demonstrating the usage of the timing module."""

import sys
from pathlib import Path
from time import sleep

python_dir = Path(__file__).parent.parent
sys.path.append(str(python_dir / 'src'))
sys.path.append(str(python_dir.parent / 'build' / 'binding'))

from demo.timing import (  # noqa: E402
    Timer,
    benchmark,
    measure_time,
    time_function,
    to_human_readable,
)


def demonstrate_basic_timer() -> None:
    """Demonstrate basic timer usage with manual start / stop."""
    print('\n1. Basic Timer Usage:')

    timer = Timer()
    print('   Timer started...')

    sleep(0.05)

    timer.stop()
    print(f'   Timer stopped after {timer.elapsed_str}')
    print(f'   Elapsed nanoseconds: {timer.elapsed_ns}')
    print(f'   Elapsed microseconds: {timer.elapsed_us}')
    print(f'   Elapsed milliseconds: {timer.elapsed_ms}')

    timer.reset()
    print('   Timer reset')

    sleep(0.025)
    print(f'   New elapsed time: {timer.elapsed_str}')


def demonstrate_scoped_timer() -> None:
    """Demonstrate context-manager-based scoped timer."""
    print('\n2. Scoped Timer Usage:')

    with measure_time('Scoped operation') as timer:
        sleep(0.03)
        print('   Doing some work inside scoped timer...')

    # You can also access the timer after the scope ends.
    print(f'   Elapsed time: {timer.elapsed_str}')


def demonstrate_benchmarking() -> None:
    """Demonstrate benchmarking."""
    print('\n3. Benchmarking:')

    def simple_task() -> int:
        total = 0
        for i in range(1000):
            total += i
        return total

    result = benchmark('Simple task', simple_task, iterations=1000)
    result.print()

    def list_ops() -> None:
        data = list(range(1000))
        data.sort(reverse=True)

    list_result = benchmark('List Operations', list_ops, iterations=500)
    list_result.print()


def demonstrate_utility_functions() -> None:
    """Demonstrate utility functions."""
    print('\n4. Utility Functions:')

    elapsed = time_function(lambda: sleep(0.0005))
    elapsed_str = to_human_readable(elapsed)
    print(f'   time_function result: {elapsed}ns ({elapsed_str})')


def main() -> None:
    """Run all timing examples."""
    print('=== Timing Module Example ===')

    demonstrate_basic_timer()
    demonstrate_scoped_timer()
    demonstrate_benchmarking()
    demonstrate_utility_functions()

    print('\n=== Timing Module Example Completed ===')


if __name__ == '__main__':
    main()

File: examples/timing_example.cpp
/**
 * @file timing_example.cpp
 * @brief Example demonstrating the usage of the timing module
 */

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <functional>
#include <numeric>
#include <print>
#include <thread>
#include <vector>

#include "timing/benchmark.hpp"
#include "timing/timer.hpp"

using cpp_features::timing::BenchmarkRunner;
using cpp_features::timing::ProfileFunction;
using cpp_features::timing::ScopedTimer;
using cpp_features::timing::TimeFunction;
using cpp_features::timing::Timer;

namespace {

/**
 * @brief Demonstrate basic timer usage with manual start / stop
 */
void DemonstrateBasicTimer() {
  std::println("\n1. Basic Timer Usage:");

  Timer timer;
  std::println("   Timer started...");

  std::this_thread::sleep_for(std::chrono::milliseconds(50));

  timer.Stop();
  std::println("   Timer stopped after {}", timer.GetElapsedString());
  std::println("   Elapsed nanoseconds: {}", timer.GetElapsed<std::chrono::nanoseconds>());
  std::println("   Elapsed microseconds: {}", timer.GetElapsed<std::chrono::microseconds>());
  std::println("   Elapsed milliseconds: {}", timer.GetElapsed<std::chrono::milliseconds>());

  timer.Reset();
  std::println("   Timer reset");

  std::this_thread::sleep_for(std::chrono::milliseconds(25));
  std::println("   New elapsed time: {}", timer.GetElapsedString());
}

/**
 * @brief Demonstrate RAII-based scoped timer
 */
void DemonstrateScopedTimer() {
  std::println("\n2. Scoped Timer Usage:");

  {
    ScopedTimer timer{"Scoped operation"};
    std::this_thread::sleep_for(std::chrono::milliseconds(30));
    std::println("   Doing some work inside scoped timer...");
  }

  {
    ScopedTimer timer{
        "Scoped operation with callback",
        [](std::int64_t ns) { std::println("   Callback received: {}ns", ns); },
    };
    std::this_thread::sleep_for(std::chrono::milliseconds(20));
    std::println("   Doing some work with callback...");
  }
}

/**
 * @brief Demonstrate benchmarking capabilities
 */
void DemonstrateBenchmarking() {
  std::println("\n3. Benchmarking:");

  auto simple_task = []() {
    volatile int sum = 0;
    for (int i = 0; i < 1000; ++i) {
      sum += i;
    }
  };

  auto result = BenchmarkRunner::Benchmark("Simple task", simple_task, 1000);
  BenchmarkRunner::PrintResult(result);

  auto vector_ops = []() {
    std::vector<int> vec(1000);
    // NOLINTNEXTLINE(modernize-use-ranges): Not supported yet
    std::iota(vec.begin(), vec.end(), 1);
    std::ranges::sort(vec, std::greater<>());
  };

  auto vector_result = BenchmarkRunner::Benchmark("Vector Operations", vector_ops, 500);
  BenchmarkRunner::PrintResult(vector_result);
}

/**
 * @brief Demonstrate utility functions and chrono features
 */
void DemonstrateUtilityFunctions() {
  std::println("\n4. Utility Functions:");

  // TimeFunction utility
  auto result = TimeFunction([]() { std::this_thread::sleep_for(std::chrono::microseconds(500)); });
  std::println("   TimeFunction result: {}ns", result);

  // ProfileFunction with single execution
  ProfileFunction("Profile test", []() {
    volatile double sum = 0.0;
    for (int i = 0; i < 10000; ++i) {
      sum += std::sqrt(i);
    }
  });

  // ProfileFunction with benchmarking
  ProfileFunction(
      "Profile benchmark",
      []() {
        std::vector<int> data(1000);
        // NOLINTNEXTLINE(modernize-use-ranges): Not supported yet
        std::iota(data.begin(), data.end(), 1);
        std::ranges::reverse(data);
      },
      100);
}

/**
 * @brief Demonstrate chrono features and duration types
 */
void DemonstrateChronoFeatures() {
  std::println("\n5. Chrono Features:");

  auto start = std::chrono::high_resolution_clock::now();
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
  auto end = std::chrono::high_resolution_clock::now();

  auto duration = end - start;
  std::println("   Direct duration measurement: {}ns",
               std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count());

  Timer timer;
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
  auto duration_obj = timer.GetElapsedDuration<std::chrono::nanoseconds>();
  std::println("   Duration object: {}ns", duration_obj.count());

  if (duration_obj > std::chrono::milliseconds(1)) {
    std::println("   Duration comparison: Operation took longer than 1ms");
  }
}

}  // namespace

auto main() -> int {
  std::println("=== Timing Module Example ===");

  DemonstrateBasicTimer();
  DemonstrateScopedTimer();
  DemonstrateBenchmarking();
  DemonstrateUtilityFunctions();
  DemonstrateChronoFeatures();

  std::println("\n=== Timing Module Example Completed ===");
  return 0;
}
